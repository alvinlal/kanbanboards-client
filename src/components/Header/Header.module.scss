header {
  display: flex;
  align-items: center;
  justify-content: space-between;
  height: 88px;
  width: 100%;
  background-color: $primary-300;
  padding: 0px 24px;
  font-size: 19px;
  position: relative;

  .logo__title {
    display: flex;
    align-items: center;
    color: white;
    h3 {
      font-weight: $bold;
      margin-left: 16px;
    }

    .hamburger__bars {
      display: none;
    }
  }

  .header__search {
    position: absolute;
    left: 50%;
    transform: translateX(-50%);

    height: auto;
  }

  nav {
    display: flex;
    align-items: center;
    color: white;

    a {
      margin-left: 16px;
      text-decoration: none;
      font-weight: $bold;
      color: white;
      white-space: nowrap;

      &:hover {
        border-bottom: 3px solid white;
      }
    }

    button {
      text-decoration: none;
      background-color: transparent;
      font-size: 19px;
      font-weight: $bold;
      color: white;
      border: none;
      cursor: pointer;
      margin-left: 16px;
      &:hover {
        border-bottom: 3px solid white;
      }
    }
  }

  @media screen and (max-width: $sm) {
    height: 64px;
    padding: 0px 16px;
    font-size: 16px;

    .logo__title {
      .logo {
        display: none;
      }
      .hamburger__bars {
        display: block;
        transform: scaley(1.7);
      }
    }

    nav {
      .logout__btn {
        display: none;
      }
      button {
        font-size: 16px;
      }
    }
    h3 {
      display: none;
    }
    .header__search {
      position: static;
      transform: none;
      width: 100%;
      margin-left: 8px;
    }
  }
}
