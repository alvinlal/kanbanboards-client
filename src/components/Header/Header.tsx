import { Link } from 'react-router-dom';
import { Bars3Icon } from '@heroicons/react/24/outline';
import styles from './Header.module.scss';
import { ReactComponent as Logo } from '../../assets/icons/logo.svg';
import Search from '../Search/Search';

interface HeaderProps extends React.HTMLAttributes<HTMLElement> {
  authenticated: boolean;
  boardPage: boolean;
}

const Header: React.FC<HeaderProps> = ({
  authenticated,
  boardPage,
  ...restProps
}) => {
  return (
    <header className={styles.header} {...restProps}>
      <div className={styles.logo__title}>
        <Logo height={48} width={48} className={boardPage ? styles.logo : ''} />
        {boardPage && (
          <Bars3Icon
            height={48}
            width={38}
            className={styles.hamburger__bars}
          />
        )}
        <h3>KanbanBoards</h3>
      </div>
      <div className={styles.header__search}>{boardPage && <Search />}</div>
      <nav>
        {authenticated ? (
          <>
            {!boardPage && <Link to="/allboards">Your Boards</Link>}
            <button
              className={boardPage ? styles.logout__btn : ''}
              onClick={() => {}}
              type="button"
            >
              Logout
            </button>
          </>
        ) : (
          <>
            <Link to="/auth/signup">Signup</Link>
            <Link to="/auth/login">Login</Link>
          </>
        )}
      </nav>
    </header>
  );
};

export default Header;
