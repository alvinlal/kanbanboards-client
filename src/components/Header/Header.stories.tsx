import { ComponentStory, ComponentMeta } from '@storybook/react';
import { MemoryRouter } from 'react-router-dom';
import Header from './Header';

export default {
  title: 'Components/Header',
  component: Header,
} as ComponentMeta<typeof Header>;

const Template: ComponentStory<typeof Header> = (args) => (
  <MemoryRouter>
    <Header {...args} />
  </MemoryRouter>
);

export const BoardPage = Template.bind({});

BoardPage.storyName = 'In Board Page';

BoardPage.args = {
  authenticated: true,
  boardPage: true,
};

export const OtherPagesAuthenticated = Template.bind({});
OtherPagesAuthenticated.storyName = 'Authenticated user in all other pages';

OtherPagesAuthenticated.args = {
  authenticated: true,
  boardPage: false,
};

export const OtherPageUnAuthenticated = Template.bind({});
OtherPageUnAuthenticated.storyName = 'UnAuthenticated user in all other pages';

OtherPageUnAuthenticated.args = {
  authenticated: false,
  boardPage: false,
};
