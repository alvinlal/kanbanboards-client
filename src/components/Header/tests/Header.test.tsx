/**
 * Requirements:-
 * 1. Should show only signup and login links while unauthenticated
 * 2. Should show only your boards and logout links while authenticated
 * 3. Should show only logout link and search bar while on board page
 * 4. Should open nav while clicking on hamburger menu on mobile
 * 5. Should match snapshot for authenticated user
 * 6. Should match snapshot for unauthenticated user
 * 7. Should match snapshot for board page
 */

import { render, screen } from '@testing-library/react';
import { BrowserRouter, MemoryRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';
import Header from '../Header';

describe('Header.tsx', () => {
  const getHeaderLinks = () => {
    const signUpLink = screen.queryByRole('link', {
      name: 'Signup',
    });
    const loginLink = screen.queryByRole('link', {
      name: 'Login',
    });
    const logoutLink = screen.queryByRole('button', {
      name: 'Logout',
    });
    const yourBoardsLink = screen.queryByRole('link', {
      name: 'Your Boards',
    });

    return {
      signUpLink,
      loginLink,
      logoutLink,
      yourBoardsLink,
    };
  };

  it('Should show only signup and login links while unauthenticated', () => {
    render(<Header authenticated={false} boardPage={false} />, {
      wrapper: BrowserRouter,
    });

    const { signUpLink, loginLink, logoutLink, yourBoardsLink } = getHeaderLinks();
    const searchBar = screen.queryByRole('searchbox');

    expect(signUpLink).toBeInTheDocument();
    expect(loginLink).toBeInTheDocument();
    expect(logoutLink).not.toBeInTheDocument();
    expect(yourBoardsLink).not.toBeInTheDocument();
    expect(searchBar).not.toBeInTheDocument();
  });

  it('Should show only your boards and logout links while authenticated', () => {
    render(<Header authenticated boardPage={false} />, {
      wrapper: BrowserRouter,
    });

    const { signUpLink, loginLink, logoutLink, yourBoardsLink } = getHeaderLinks();
    const searchBar = screen.queryByRole('searchbox');
    expect(logoutLink).toBeInTheDocument();
    expect(yourBoardsLink).toBeInTheDocument();
    expect(signUpLink).not.toBeInTheDocument();
    expect(loginLink).not.toBeInTheDocument();
    expect(searchBar).not.toBeInTheDocument();
  });

  it('Should show only logout link and search bar while on board page', () => {
    render(<Header authenticated boardPage />, {
      wrapper: BrowserRouter,
    });

    const { signUpLink, loginLink, logoutLink, yourBoardsLink } = getHeaderLinks();
    const searchBar = screen.queryByRole('searchbox');

    expect(logoutLink).toBeInTheDocument();
    expect(searchBar).toBeInTheDocument();
    expect(logoutLink).toBeInTheDocument();
    expect(signUpLink).not.toBeInTheDocument();
    expect(loginLink).not.toBeInTheDocument();
    expect(yourBoardsLink).not.toBeInTheDocument();
  });

  it.todo('Should open nav while clicking on hamburger menu on mobile');

  it('Should match snapshot for authenticated user', () => {
    const tree = renderer.create(
      <MemoryRouter>
        <Header authenticated boardPage={false} />
      </MemoryRouter>
    );
    expect(tree).toMatchSnapshot();
  });

  it('Should match snapshot for unauthenticated user', () => {
    const tree = renderer.create(
      <MemoryRouter>
        <Header authenticated={false} boardPage={false} />
      </MemoryRouter>
    );
    expect(tree).toMatchSnapshot();
  });

  it('Should match snapshot for board page', () => {
    const tree = renderer.create(
      <MemoryRouter>
        <Header authenticated boardPage />
      </MemoryRouter>
    );
    expect(tree).toMatchSnapshot();
  });
});
