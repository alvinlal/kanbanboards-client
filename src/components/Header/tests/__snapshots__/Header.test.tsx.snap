// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header.tsx Should match snapshot for authenticated user 1`] = `
<header
  className="header"
>
  <div
    className="logo__title"
  >
    <div
      className=""
      cursor="pointer"
      height={42}
      onClick={[Function]}
      width={42}
    />
    <h4>
      KanbanBoards
    </h4>
  </div>
  <div
    className="header__search"
  />
  <nav>
    <a
      href="/allboards"
      onClick={[Function]}
    >
      Your Boards
    </a>
    <button
      className=""
      onClick={[Function]}
      type="button"
    >
      Logout
    </button>
  </nav>
</header>
`;

exports[`Header.tsx Should match snapshot for board page 1`] = `
<header
  className="header"
>
  <div
    className="logo__title"
  >
    <div
      className="logo"
      cursor="pointer"
      height={42}
      onClick={[Function]}
      width={42}
    />
    <svg
      aria-hidden="true"
      className="hamburger__bars"
      fill="none"
      height={48}
      stroke="currentColor"
      strokeWidth={1.5}
      viewBox="0 0 24 24"
      width={38}
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M3.75 6.75h16.5M3.75 12h16.5m-16.5 5.25h16.5"
        strokeLinecap="round"
        strokeLinejoin="round"
      />
    </svg>
    <h4>
      KanbanBoards
    </h4>
  </div>
  <div
    className="header__search"
  >
    <div
      className="search"
    >
      <div
        className="search__bar"
      >
        <div
          className="search__input"
        >
          <svg
            aria-hidden="true"
            fill="none"
            height={32}
            stroke="currentColor"
            strokeWidth={3}
            viewBox="0 0 24 24"
            width={32}
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M21 21l-5.197-5.197m0 0A7.5 7.5 0 105.196 5.196a7.5 7.5 0 0010.607 10.607z"
              strokeLinecap="round"
              strokeLinejoin="round"
            />
          </svg>
          <input
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            placeholder="Search"
            type="search"
          />
        </div>
      </div>
    </div>
  </div>
  <nav>
    <button
      className="logout__btn"
      onClick={[Function]}
      type="button"
    >
      Logout
    </button>
  </nav>
</header>
`;

exports[`Header.tsx Should match snapshot for unauthenticated user 1`] = `
<header
  className="header"
>
  <div
    className="logo__title"
  >
    <div
      className=""
      cursor="pointer"
      height={42}
      onClick={[Function]}
      width={42}
    />
    <h4>
      KanbanBoards
    </h4>
  </div>
  <div
    className="header__search"
  />
  <nav>
    <a
      href="/signup"
      onClick={[Function]}
      tabIndex={0}
    >
      Signup
    </a>
    <a
      href="/login"
      onClick={[Function]}
    >
      Login
    </a>
  </nav>
</header>
`;
